### PR TITLE
MetricCountSeqAndTokens counst tokens in label (if exist) + improving epoch metrics print

### PR DESCRIPTION
**Pull request by michalozeryflato**
1. add optional decoder_input to MetricCountSeqAndTokens
2. print traceback before throw - helps in debugging
3. modify epoch stats printing - each column has its own (fixed) width

### CODE REVIEW

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 1]
```diff
@@ -82,12 +85,15 @@ def get_state(self) -> dict:
...
     encoder_input_key: str,
+    decoder_input_key: Optional[str] = None,
     ignore_index: Optional[int] = None,
 ) -> Dict[str, torch.Tensor]:
     """Count number of sequences and tokens
     Args:
         encoder_input_key:
             key to encoder_input
+        decoder_input_key:
+            key to encoder_input
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:96]: key to *decoder_input :)

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 2]
```diff
@@ -98,19 +104,29 @@ def _count_seq_and_tokens_update(
...
     seq_num = torch.tensor(
         mask.shape[0], dtype=torch.int64, device=encoder_input.device
     )
     token_num = mask.sum().to(dtype=torch.int64)
 
+    if decoder_input_key is not None and decoder_input_key in batch_dict:
+        decoder_input = batch_dict[decoder_input_key].detach()
+        mask2 = get_mask(decoder_input, ignore_index)
+        assert mask2.shape[0] == mask.shape[0]
+        token_num += mask2.sum().to(dtype=torch.int64)
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:118]: Just making sure:
Here's the support for counting the labels tokens as well?

If so, best to double check with @mosheraboh :)

Comment (michalozeryflato) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:118]: @SagiPolaczek I updated @mosheraboh, but better confirm the change with him.

File: fuse/utils/misc/misc.py [hunk 3]
```diff
@@ -158,22 +158,22 @@ def time_display(seconds: int, granularity: int = 3) -> str:
...
-    max_val_width = np.vectorize(len)(
-        df.values.astype(str)
-    ).max()  # get maximum length of all values
-    max_col_width = max(
-        [len(x) for x in df.columns]
-    )  # get the maximum lengths of all column names
-    col_width = max(max_col_width, max_val_width, col_width)
-
-    dashes = (col_width + 2) * len(df.columns.values)
+    max_val_width_per_col = np.vectorize(len)(df.values.astype(str)).max(
```
Comment (mosheraboh) [inline fuse/utils/misc/misc.py:161]: Here you've modified the code to set the width per column instead of a single width for all columns?

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 1 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:96]: ☝️

**Review by mosheraboh — APPROVED** [review]
LGTM

File: fuse/utils/misc/misc.py (hunk 3 above)
Comment (michalozeryflato) [reply fuse/utils/misc/misc.py:161]: yes, to make the overall width smaller - so it better fits the screen

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 4]
```diff
@@ -177,18 +194,18 @@ def _perplexity_update(
     preds = preds.detach()
     target = target.detach()
 
-    preds = preds.reshape(-1, preds.shape[-1])
-    target = target.reshape(-1)
+    preds = preds.view(-1, preds.shape[-1])
```
Comment (mosheraboh) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:197]: why? to save memory?

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 5]
```diff
@@ -137,6 +153,7 @@ def __init__(
 
 # Copied internal function https://github.com/Lightning-AI/metrics/blob/825d17f32ee0b9a2a8024c89d4a09863d7eb45c3/src/torchmetrics/functional/text/perplexity.py#L68
 # copied and not imported to not be affected by internal interface modifications.
+# modifications: (1) reshape => view (2) apply mask at the beginning of computation (3) use torch.gather
```
Comment (mosheraboh) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:156]: Nice solution.

**Review by mosheraboh — APPROVED** [review]
Looks great

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 4 above)
Comment (michalozeryflato) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:197]: @mosheraboh From the documentation of [reshape](https://pytorch.org/docs/stable/generated/torch.reshape.html#torch.reshape): When possible, the returned tensor will be a **view of input. Otherwise, it will be a copy**. Contiguous inputs and inputs with compatible strides can be reshaped without copying, but you should not depend on the copying vs. viewing behavior.

Question is: 
(1) do we assume view will always work, and otherwise alert the user. Or 
(2) do we want to be permissive and copy the data (despite additional GPU memory) when view is not possible (i.e. memory is not contiguous)

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 6]
```diff
@@ -177,24 +194,23 @@ def _perplexity_update(
...
-    target = target.reshape(-1)
+    preds = preds.view(-1, preds.shape[-1])
+    target = target.view(-1)
 
     if ignore_index is not None:
         mask = target.ne(ignore_index)
-        target = target.where(
-            target != ignore_index, torch.tensor(0, device=target.device)
-        )
+        target = target[mask]
```
[the remark this answers is not in the document]
Comment (SagiPolaczek) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:202]: I'm not sure it's critical (maybe it's even better, saving memory)

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 4 above)
Comment (SagiPolaczek) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:197]: Is it really that expensive in memory?

If it is, than I would warn the user conditionally to `Tensor.is_contiguous()` result :)

**Review by SagiPolaczek — DISMISSED** [review]
THANK YOU!

Two comments inline

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (michalozeryflato) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:204]: yes, I have reduced the side of pred and target, since at the end the entries corresponding to the mask are dropped. So it seems to produce the same results, but with less memory/time (not having to process the masked entries)

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 7]
```diff
@@ -177,24 +194,25 @@ def _perplexity_update(
     preds = preds.detach()
     target = target.detach()
 
+    # reshape attempts to create a view, and COPIES the data if fails.
+    # an issue to consider:  use view and alert the user if it fails?
     preds = preds.reshape(-1, preds.shape[-1])
     target = target.reshape(-1)
 
     if ignore_index is not None:
         mask = target.ne(ignore_index)
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:203]: I guess we can use `get_mask()` here

**Review by SagiPolaczek — APPROVED** [review]
LGTM! Thanks!
